crates/svm-runtime/src/import.rs

```diff
@@ -1,52 +1,201 @@
+
+                        Ok(vals)
+                    } else {
+                        Err(RuntimeError::new("Invalid WASM values"))
+                    }
+                };
+
+            let func_ty = self.wasmer_function_ty();
+
+            /// making the input `&mut Context` appear as `*const c_void`
+            let inner_env = ctx as *mut Context as *const Context as *const c_void;
+            let host_env = self.host_env;
+
+            /// The import used `env` (using Wasmer terminology) will be a struct of `svm_env_t`
+            /// This `#[repr(C)]` struct will contain two pointers to two types of `env`:
+            ///
+            /// 1. SVM inner env - a pointer to the `Context`
+            ///    Once SVM has finished executing a transaction its memory will be deallocated.
+            ///
+            /// 2. Host env - a pointer given as input by the so-called `Host`
+            ///    The responsibility of release that memory is up to the caller (the `host`).
+            let func_env = svm_env_t {
+                inner_env,
+                host_env,
+            };
+
+            /// The heap-allocated `func_env` will be deallocated by later by `SVM` running runtime.
+            /// (See method `funcs_envs_destroy` under `src/runtime/default.rs`)
+            let func_env = Box::into_raw(Box::new(func_env));
+
+            let func = Function::new_with_env(store, &func_ty, func_env, inner_callback);
+            let export = func.to_export();
+
+            let func_env = func_env as *const svm_env_t;
+
+            (export, func_env)
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn namespace(&self) -> &str {
+        &self.namespace
+    }
+
+    fn wasmer_function_ty(&self) -> FunctionType {
         let params = to_wasmer_types(&self.params);
         let returns = to_wasmer_types(&self.returns);
-        let signature = FunctionType::new(params, returns);
-
-        // TODO: needs to free `ctx`.
-        let boxed_ctx = Box::new(ctx);
-        let vmctx = Box::into_raw(boxed_ctx) as *mut _ as *mut VMContext;
-
-        let func = ExportFunction {
-            address: self.func_ptr as _,
-            vmctx,
-            signature,
-            kind: VMFunctionKind::Static,
-            call_trampoline: None,
-        };
 
-        func.into()
+        FunctionType::new(params, returns)
     }
 }
 
-fn to_wasmer_types(types: &[WasmType]) -> Vec<WasmerType> {
+fn to_wasmer_types(types: &[WasmType]) -> Vec<Type> {
     types
         .iter()
         .map(|ty| match ty {
-            WasmType::I32 => WasmerType::I32,
-            WasmType::I64 => WasmerType::I64,
+            WasmType::I32 => Type::I32,
+            WasmType::I64 => Type::I64,
             _ => panic!("Only i32 and i64 are supported."),
         })
         .collect()
 }
+
+#[inline]
+fn to_wasm_values(bytes: &svm_byte_array, types: &[WasmType]) -> Option<Vec<WasmValue>> {
```

Comment:
Wouldn't it be better to return `Result` instead of `Option`, since you later transform the `None` option to `Err` anyway? That way you could return more specific errors from `to_wasm_values`.